### ISSUE TITLE
Why must 'Data' be String or JSONArray?

### ISSUE THREAD

**Issue by JackLivio** [smartdevicelink/sdl_android#230]
In SDLProxyBase, the method [SendOnSystemRequestToUrl] (https://github.com/smartdevicelink/sdl_android/blob/hotfix/fixes_226/sdl_android_lib/src/com/smartdevicelink/proxy/SdlProxyBase.java#L782), requires the response from SDL server to have an object "data" which must be of type String or JSONArray.

[Line 892](https://github.com/smartdevicelink/sdl_android/blob/hotfix/fixes_226/sdl_android_lib/src/com/smartdevicelink/proxy/SdlProxyBase.java#L892)

```
			JSONObject jsonResponse = new JSONObject(response.toString());				
			if (jsonResponse.get("data") instanceof JSONArray) 
			{
				JSONArray jsonArray = jsonResponse.getJSONArray("data");
				for (int i=0; i<jsonArray.length(); i++) 
				{
					if (jsonArray.get(i) instanceof String) 
					{
						cloudDataReceived.add(jsonArray.getString(i));
						//Log.i("sendOnSystemRequestToUrl", "jsonArray.getString(i): " + jsonArray.getString(i));
					}
				}
			} 
			else if (jsonResponse.get("data") instanceof String) 
			{
				cloudDataReceived.add(jsonResponse.getString("data"));
				//Log.i("sendOnSystemRequestToUrl", "jsonResponse.getString(data): " + jsonResponse.getString("data"));
			} 
			else 
			{
				DebugTool.logError("sendOnSystemRequestToUrl: Data in JSON Object neither an array nor a string.");
				//Log.i("sendOnSystemRequestToUrl", "sendOnSystemRequestToUrl: Data in JSON Object neither an array nor a string.");
				return;
			}
```

Is there a reason why "Data" must be limited to these two types? 
Can support for "Data" as a JSONObject be added?

ie
```
else if (jsonResponse.get("data") instanceof JSONObject) {
...
}
```

**Comment by justinjdickow** [comment]
Is there a particular behavior in a release of SDL Core that requires `data` to be an Object?

**Comment by JackLivio** [comment]
When I was working on getting SDL Server to send a policy table update, the current implementation of android would only accept  "data" if its contents (the policy table json object) were parsed to a string or placed inside of a JsonArray.

When core receives the policy table update from the server, it writes "data" to a file and then sends the file to the HMI where it is supposed be decode/decrypt the policy table update for Core to be able to properly update the SQL database.

When using the web hmi, it is not possible to perform a decode/decryption process since you cannot write to a new file using the browser/javascript. Therefore, the current web hmi implementation will either pass back the same file path that was received by core, or return a file path with a 
"Sample" policy table that is "hardcoded" in the sdl_hmi directory (kind of a crappy hack since the update isn't actually coming from the server).

When the HMI attempts to send the same file path that was received from core, the policy table update will fail when Core complains that the file is improperly formatted. This entire issue would be resolved if the HMI was able to decode the policy update and write the result to a file.

Also, if the policy table didn't have to be cast to a string or JSON array from the very start, then core would have written "data" to a file as a properly formatted JSON object that it would later be able to read and perform policy table update.

tl;dr
Core can't read a policy table update when the server casts it as a string or array.

**Comment by justinjdickow** [comment]
@joeygrover @mrapitis bump

**Comment by mrapitis** [comment]
Including the suggested else if should not break existing functionality, however I am still not clear why sdlcore requires the data to be an object.

**Comment by justinjdickow** [comment]
@JackLivio Maybe there is something about the sdl_server implementation that needs to change to conform to the existing structure on Mobile?

**Comment by justinjdickow** [comment]
@JackLivio Can you answer the above question? We need to confirm our implementation on SDL Core before it's merged, this info might help

**Comment by JackLivio** [comment]
Existing mobile structure does not need to change, I was able to send the data object as a JSON array and then do the parsing in Core using the json.cpp lib. 

I still think it would be convenient to add the extra if statement in mobile allowing for JSON objects. Not sure why its limited to types string and JSON Array.
